syntax = "proto3";

package api.room;

import "google/protobuf/timestamp.proto";

service Room {
  // Unary RPCs
  rpc CreateRoom(CreateRoomRequest) returns (CreateRoomResponse);
  rpc GetRoom(GetRoomRequest) returns (GetRoomResponse);
  rpc ListRooms(ListRoomsRequest) returns (ListRoomsResponse);
  rpc LoadHistory(LoadHistoryRequest) returns (LoadHistoryResponse);

  // Bidirectional streaming - main session
  // One stream per room per user. JoinRoom must be the first ClientMessage.
  rpc RoomSession(stream ClientMessage) returns (stream ServerEvent);
}

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

enum ParticipantType {
  PARTICIPANT_TYPE_UNSPECIFIED = 0;
  HUMAN = 1;
  LLM = 2;
}

enum Role {
  ROLE_UNSPECIFIED = 0;
  ADMIN = 1;
  MEMBER = 2;
  VIEWER = 3;
}

enum PollStatus {
  POLL_STATUS_UNSPECIFIED = 0;
  POLL_OPEN = 1;
  POLL_CLOSED = 2;
}

enum ChatStyle {
  CHAT_STYLE_UNSPECIFIED = 0;  // default behavior
  CHAT_STYLE_CONVERSATIONAL = 1;  // short, punchy responses
  CHAT_STYLE_DETAILED = 2;  // thorough explanations
  CHAT_STYLE_BULLET = 3;  // structured lists
}

enum RoomVisibility {
  ROOM_VISIBILITY_UNSPECIFIED = 0;  // defaults to public
  ROOM_VISIBILITY_PUBLIC = 1;       // anyone can see and join
  ROOM_VISIBILITY_PRIVATE = 2;      // only creator sees in list, link sharing still works
}

// ---------------------------------------------------------------------------
// Shared messages
// ---------------------------------------------------------------------------

message LLMConfig {
  string id = 1;              // stable identifier, e.g. "claude-1"
  string model = 2;           // OpenRouter model id
  string persona = 3;         // system prompt / role description
  string display_name = 4;    // human-readable, e.g. "Claude"
  string title = 5;           // job title, e.g. "VP Engineering"
  ChatStyle chat_style = 6;   // response style (conversational, detailed, bullet)
  string avatar = 7;          // emoji avatar, e.g. "ü§ñ" or "üß†"
}

message Participant {
  string id = 1;
  string name = 2;
  Role role = 3;
  ParticipantType type = 4;
  string title = 5;
  bool is_online = 6;  // Whether participant is currently connected
  string avatar = 7;   // emoji avatar, e.g. "üë§" or "üßë‚Äçüíª"
}

message RoomInfo {
  string room_id = 1;
  string name = 2;
  google.protobuf.Timestamp created_at = 3;
  string created_by = 4;
  repeated LLMConfig llms = 5;
  string description = 6;  // room-wide context (e.g. "Fintech startup meeting room")
  RoomVisibility visibility = 7;  // public or private
}

message Message {
  string message_id = 1;
  string sender_id = 2;
  string sender_name = 3;
  ParticipantType sender_type = 4;
  string content = 5;
  optional string reply_to = 6;
  google.protobuf.Timestamp timestamp = 7;
  optional string poll_id = 8;  // If set, this message is a poll (content = question)
}

// ---------------------------------------------------------------------------
// Polls
// ---------------------------------------------------------------------------

message PollVote {
  string voter_id = 1;
  string voter_name = 2;
  string reason = 3;  // optional reasoning (LLMs often explain their vote)
  google.protobuf.Timestamp voted_at = 4;
}

message PollOption {
  string id = 1;
  string text = 2;
  string description = 3;  // optional elaboration
  repeated PollVote votes = 4;
}

message Poll {
  string poll_id = 1;
  string room_id = 2;
  string creator_id = 3;
  string creator_name = 4;
  ParticipantType creator_type = 5;
  string question = 6;
  repeated PollOption options = 7;
  bool allow_multiple = 8;
  bool anonymous = 9;
  PollStatus status = 10;
  google.protobuf.Timestamp created_at = 11;
  google.protobuf.Timestamp closed_at = 12;
  bool mandatory = 13;  // If true, LLMs must vote (can't opt out)
}

// ---------------------------------------------------------------------------
// Unary: CreateRoom
// ---------------------------------------------------------------------------

message CreateRoomRequest {
  string name = 1;
  repeated LLMConfig llms = 2;
  // created_by should come from auth context in production;
  // passed explicitly for now while auth is not implemented.
  string created_by = 3;
  string description = 4;  // room-wide context for all LLMs
  RoomVisibility visibility = 5;  // public (default) or private
}

message CreateRoomResponse {
  string room_id = 1;
}

// ---------------------------------------------------------------------------
// Unary: GetRoom
// ---------------------------------------------------------------------------

message GetRoomRequest {
  string room_id = 1;
}

message GetRoomResponse {
  RoomInfo room = 1;
  repeated Participant participants = 2;  // currently online
}

// ---------------------------------------------------------------------------
// Unary: ListRooms
// ---------------------------------------------------------------------------

message ListRoomsRequest {
  // Filter to rooms this user has joined. Empty = all rooms.
  optional string user_id = 1;
  uint32 limit = 2;                     // default 20
  optional string cursor = 3;           // opaque pagination token
}

message ListRoomsResponse {
  repeated RoomInfo rooms = 1;
  optional string next_cursor = 2;
}

// ---------------------------------------------------------------------------
// Unary: LoadHistory
// ---------------------------------------------------------------------------

message LoadHistoryRequest {
  string room_id = 1;
  uint32 limit = 2;                     // default 50
  optional string cursor = 3;           // opaque; MSG# sort key
}

message LoadHistoryResponse {
  repeated Message messages = 1;
  optional string next_cursor = 2;      // null = no more history
}

// ---------------------------------------------------------------------------
// Bidi Stream: Client ‚Üí Server
// ---------------------------------------------------------------------------

message ClientMessage {
  oneof payload {
    JoinRoom join = 1;
    SendMessage message = 2;
    TypingIndicator typing = 3;
    InterruptLLM interrupt = 4;
    Ping ping = 5;
    AddLLM add_llm = 6;
    UpdateLLM update_llm = 7;
    CreatePoll create_poll = 8;
    CastVote cast_vote = 9;
    ClosePoll close_poll = 10;
    RemoveLLM remove_llm = 11;
    UpdateRoomDescription update_room_description = 12;
  }
}

message UpdateRoomDescription {
  string description = 1;
}

message CreatePoll {
  string question = 1;
  repeated PollOptionInput options = 2;
  bool allow_multiple = 3;
  bool anonymous = 4;
  bool mandatory = 5;  // If true, LLMs must vote
}

message PollOptionInput {
  string text = 1;
  string description = 2;  // optional
}

message CastVote {
  string poll_id = 1;
  repeated string option_ids = 2;  // supports multiple if poll allows
  string reason = 3;  // optional reasoning
}

message ClosePoll {
  string poll_id = 1;
}

message AddLLM {
  LLMConfig llm = 1;
}

message UpdateLLM {
  string llm_id = 1;
  optional string model = 2;
  optional string persona = 3;
  optional string display_name = 4;
  optional string title = 5;
  optional ChatStyle chat_style = 6;
  optional string avatar = 7;
}

message RemoveLLM {
  string llm_id = 1;
}

message JoinRoom {
  string room_id = 1;
  // user_id is client-generated for anonymous users.
  // Server should validate / assign via auth when available.
  string user_id = 2;
  string display_name = 3;
  Role role = 4;
  string title = 5;
  string avatar = 6;  // emoji avatar
}

message SendMessage {
  string content = 1;
  // Client-parsed @mentions. Server re-validates against room LLM configs.
  repeated string mentions = 2;
  optional string reply_to = 3;
}

message TypingIndicator {
  bool is_typing = 1;
}

message InterruptLLM {
  string llm_id = 1;
  // Which response to interrupt (required when LLM has concurrent responses)
  optional string message_id = 2;
}

message Ping {}

// ---------------------------------------------------------------------------
// Bidi Stream: Server ‚Üí Client
// ---------------------------------------------------------------------------

message ServerEvent {
  oneof payload {
    RoomState room_state = 1;
    MessageReceived message_received = 2;
    UserJoined user_joined = 3;
    UserLeft user_left = 4;
    LLMThinking llm_thinking = 5;
    LLMChunk llm_chunk = 6;
    LLMDone llm_done = 7;
    UserTyping user_typing = 8;
    Error error = 9;
    Pong pong = 10;
    LLMAdded llm_added = 11;
    LLMUpdated llm_updated = 12;
    PollCreated poll_created = 13;
    PollVoted poll_voted = 14;
    PollClosed poll_closed = 15;
    LLMRemoved llm_removed = 16;
    RoomUpdated room_updated = 17;
  }
}

message RoomUpdated {
  RoomInfo room = 1;
}

message LLMRemoved {
  string llm_id = 1;
}

message PollCreated {
  Poll poll = 1;
}

message PollVoted {
  string poll_id = 1;
  string option_id = 2;
  PollVote vote = 3;
}

message PollClosed {
  string poll_id = 1;
  string closed_by_id = 2;
  string closed_by_name = 3;
}

message LLMAdded {
  LLMConfig llm = 1;
}

message LLMUpdated {
  LLMConfig llm = 1;  // full updated config
}

message RoomState {
  RoomInfo room = 1;
  repeated Participant participants = 2;
  repeated Message messages = 3;         // last N on join
  repeated Poll polls = 4;               // active polls in room
}

message MessageReceived {
  Message message = 1;
}

message UserJoined {
  Participant user = 1;
}

message UserLeft {
  string user_id = 1;
}

message LLMThinking {
  string llm_id = 1;
  string reply_to = 2;
}

message LLMChunk {
  string message_id = 1;
  string llm_id = 2;
  string content = 3;      // delta
  string reply_to = 4;
}

message LLMDone {
  string message_id = 1;
  string llm_id = 2;
}

message UserTyping {
  string user_id = 1;
  string user_name = 2;
  bool is_typing = 3;
}

message Error {
  string code = 1;
  string message = 2;
}

message Pong {}
